### PR TITLE
Add a Web view

### PR DESCRIPTION
**Pull request by apostergiou**
With this we add a web view for the service for monitoring the available jobs.

The feature introduces:
- an `index` page to view all available jobs and relevant information such as start time and state
- a `show` page for each where we can view the job result information alongside the job build logs. The show page provides a web tailf like feature for monitoring the build log as it unfolds.

### CODE REVIEW

File: tail_reader.go [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+)
+
+type tailReader struct {
+	io.ReadCloser
+}
+
+// Read provides a tailf like generator.
+//
+// The reader sleeps on EOF and is shutdown when the file is closed.
+// Implements the io.Reader inerface.
```
Comment (agis) [inline tail_reader.go]: This line belongs on the struct documentation.

File: tail_reader.go [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+		if n > 0 {
+			return n, nil
+		} else if err != io.EOF {
+			return n, err
+		}
+		time.Sleep(5 * time.Second)
+	}
+}
+
+func newTailReader(fileName string) (tailReader, error) {
```
Comment (agis) [inline tail_reader.go]: It's actually a `path`, not ` fileName`.

Comment (agis) [inline tail_reader.go]: It's idiomatic to return a pointer to the struct instead.

File: testdata/projects/simple-slow/docker-entrypoint.sh [hunk 3]
```diff
@@ -0,0 +1,7 @@
+#!/bin/bash
+set -e
+
+sleep 30
+touch artifacts/out.txt
+
+exit 0
```
Comment (agis) [inline testdata/projects/simple-slow/docker-entrypoint.sh]: This directory should be here only if it's actually used in mistry's tests.

File: types/build_result.go [hunk 4]
```diff
@@ -23,6 +23,9 @@ type BuildResult struct {
 
 	// The method by which the build artifacts can be fetched.
 	TransportMethod TransportMethod
+
+	// The job's start time
+	StartTime string
```
Comment (agis) [inline types/build_result.go]: This should be a `time.Time`.

File: types/job_info.go [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+)
+
+type JobInfo struct {
+	ID        string
+	Project   string
+	StartedAt time.Time
+	Output    string
+	Log       string
+	State     string
+}
```
Comment (agis) [inline types/job_info.go]: I see many duplicated fields between this and `Job`. Can we merge this into `Job`? It seems redundant.

File: server.go [hunk 6]
```diff
@@ -22,6 +45,25 @@ type Server struct {
...
+
+	// Client connections registry.
+	//
+	// Cs sent to the newClients channel are being added to the registry.
+	// We use a reference to the C struct so that the conncetions can be
+	// uniquely identified for deletions.
+	clients map[*C]bool
+
+	// A connection queue used to track all open connections by their id.
+	cq map[string]bool
```
Comment (agis) [inline server.go]: All these fields and methods could go into a separate struct, ideally extracted into its own package and exporting its http handlers that one can use selectively.

File: tail_reader.go [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+// Implements the io.Reader inerface.
+func (t tailReader) Read(b []byte) (int, error) {
+	for {
+		n, err := t.ReadCloser.Read(b)
+		if n > 0 {
+			return n, nil
+		} else if err != io.EOF {
+			return n, err
+		}
+		time.Sleep(5 * time.Second)
```
Comment (agis) [inline tail_reader.go]: Can you elaborate on why is this sleep needed?

File: job.go [hunk 8]
```diff
@@ -192,3 +192,16 @@ func (j *Job) String() string {
 		"{project=%s params=%s group=%s id=%s}",
 		j.Project, j.Params, j.Group, j.ID[:7])
 }
+
+func jobPath(buildPath, project, id string) (string, string, error) {
```
[the remark this answers is not in the document]
Comment (apostergiou) [reply job.go]: Updated.

File: job.go [hunk 9]
```diff
@@ -192,3 +192,16 @@ func (j *Job) String() string {
...
+	pPath := filepath.Join(buildPath, project, "pending", id)
+	rPath := filepath.Join(buildPath, project, "ready", id)
+	if _, err := os.Stat(pPath); err == nil {
```
[the remark this answers is not in the document]
Comment (apostergiou) [reply job.go]: Fixed.

File: local_assets.go [hunk 10]
```diff
@@ -0,0 +1,7 @@
+package main
+
+import "net/http"
+
+func staticFs() (http.FileSystem, error) {
+	return http.Dir("public"), nil
```
[the remark this answers is not in the document]
Comment (apostergiou) [reply local_assets.go]: True, deleted.

File: public/index.html [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+      </div>
+
+      <div class="large-12 cell" id="js-jobs">
+      </div>
+    </div>
+  </div>
+
+  <script src="https://unpkg.com/babel-standalone@6/babel.min.js"></script>
+  <script src="https://unpkg.com/react@16/umd/react.production.min.js"></script>
+  <script src="https://unpkg.com/react-dom@16/umd/react-dom.production.min.js"></script>
```
[the remark this answers is not in the document]
Comment (apostergiou) [reply public/index.html]: Good catch, moved inside the head tag.

File: public/js/index.js [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+              </tr>
+            )
+           })}
+        </tbody>
+      </table>
+    );
+  }
+}
+
+ReactDOM.render(<Jobs every={5000} />, JobsRoot)
```
[the remark this answers is not in the document]
Comment (apostergiou) [reply public/js/index.js]: Ok, let's try with 3sec.

File: server.go [hunk 13]
```diff
@@ -1,17 +1,40 @@
...
+	"strings"
 
 	"github.com/skroutz/mistry/types"
 )
 
+// C represents the connections to the job show page.
+//
+// Every struct instance will keep an open connection with the server
+// and listens for server side events.
+type C struct {
```
[the remark this answers is not in the document]
Comment (apostergiou) [reply server.go]: Updated.

File: tail_reader.go [hunk 14]
```diff
@@ -0,0 +1,39 @@
+package main
+
+import (
+	"io"
+	"os"
+	"time"
+)
+
+type tailReader struct {
```
[the remark this answers is not in the document]
Comment (apostergiou) [reply tail_reader.go]: Done :+1: .

File: tail_reader.go (hunk 1 above)
Comment (apostergiou) [reply tail_reader.go]: Fixed.

File: tail_reader.go (hunk 2 above)
Comment (apostergiou) [reply tail_reader.go]: true, changed.

File: testdata/projects/simple-slow/docker-entrypoint.sh (hunk 3 above)
Comment (apostergiou) [reply testdata/projects/simple-slow/docker-entrypoint.sh]: Sure, removed.

File: tail_reader.go (hunk 7 above)
Comment (apostergiou) [reply tail_reader.go]: Thanks for commenting on this. I added sleep here as a way to give up control to the scheduler while developing the feature. Nevertheless,  using it here feels like an anti-pattern and doesn't add anything so I removed it.

**Comment by agis** [comment]
This PR should also update the relevant [README section](https://github.com/skroutz/mistry/commit/f8bac8d6e6c857ec5dd2b42cf42180f5340b3193#diff-04c6e90faac2675aa89e2176d2eec7d8R19).

File: server.go (hunk 6 above)
Comment (apostergiou) [reply server.go]: Extracted into a `broker` package.

File: types/build_result.go (hunk 4 above)
Comment (apostergiou) [reply types/build_result.go]: done

File: tail_reader.go (hunk 2 above)
Comment (apostergiou) [reply tail_reader.go]: :+1:  done

File: README.md [hunk 15]
```diff
@@ -16,7 +16,7 @@ Features include:
 - caching and reusing build results
 - efficient use of disk space due to copy-on-write semantics
 - a simple JSON API for interacting with the service
-- ([wip](https://github.com/skroutz/mistry/pull/17)) a web view for inspecting the progress and result of builds
+- a web view for inspecting the progress and result of builds
```
Comment (agis) [inline README.md]: This needs a rebase. I slightly updated this point.

File: job.go [hunk 16]
```diff
@@ -207,3 +207,18 @@ func (j *Job) String() string {
 		"{project=%s params=%s group=%s id=%s}",
 		j.Project, j.Params, j.Group, j.ID[:7])
 }
+
+// State determines the job's current state by using it's path in the filesystem.
+func (j *Job) State(path, project, id string) (string, error) {
+	pPath := filepath.Join(path, project, "pending", id)
+	rPath := filepath.Join(path, project, "ready", id)
+	_, err := os.Stat(pPath)
+	if err == nil {
```
Comment (agis) [inline job.go]: What if error is non nil but also not a "not exist" error?

File: job.go [hunk 17]
```diff
@@ -207,3 +207,18 @@ func (j *Job) String() string {
...
+		return "pending", nil
+	}
+	_, err = os.Stat(rPath)
+	if err == nil {
+		return "ready", nil
+	}
+	return "", err
```
Comment (agis) [inline job.go]: I don't like the fact that we may return an empty string and no error. What would such a case indicate?

File: job.go [hunk 18]
```diff
@@ -207,3 +207,18 @@ func (j *Job) String() string {
 		"{project=%s params=%s group=%s id=%s}",
 		j.Project, j.Params, j.Group, j.ID[:7])
 }
+
+// State determines the job's current state by using it's path in the filesystem.
+func (j *Job) State(path, project, id string) (string, error) {
+	pPath := filepath.Join(path, project, "pending", id)
+	rPath := filepath.Join(path, project, "ready", id)
```
Comment (agis) [inline job.go]: Can't we use `j.PendingBuildPath` and `j.ReadyBuildPath`?

File: public/templates/show.html [hunk 19]
```diff
@@ -0,0 +1,93 @@
...
+  <meta name="viewport" content="width=device-width, initial-scale=1">
+  <title>Job Logs</title>
+  <link rel="stylesheet" href="../../css/foundation.min.css">
+</head>
+
+<body>
+
+  <div class="top-bar">
+    <div class="top-bar-left">
+      <h1><a href="/">Mistry</a></h1>
```
Comment (agis) [inline public/templates/show.html]: lowercase

File: public/templates/show.html [hunk 20]
```diff
@@ -0,0 +1,93 @@
...
+      }
+
+      const source = new EventSource('/log/{{.Project}}/{{.ID}}');
+      const jobLog = document.getElementById('js-job-log')
+      source.onmessage = function(e) {
+        jobLog.innerHTML += e.data;
+      };
+
+      source.onerror = function(e) {
+        document.body.innerHTML += "Web tail failed.";
```
Comment (agis) [inline public/templates/show.html]: Can we display more information about the error?

File: server_test.go [hunk 21]
```diff
@@ -65,3 +72,87 @@ func TestLoad(t *testing.T) {
 		<-results
 	}
 }
+
+func TestIndex(t *testing.T) {
```
Comment (agis) [inline server_test.go]: Tests functions are named according to the functions they test. This should be named `TestHandleIndex` for example.

File: server_test.go [hunk 22]
```diff
@@ -65,3 +72,87 @@ func TestLoad(t *testing.T) {
...
+	body, err := ioutil.ReadAll(result.Body)
+	if err != nil {
+		t.Fatal(err)
+	}
+	if !strings.Contains(string(body), expected) {
+		t.Errorf("Expeced body to contain %v, got %v", expected, string(body))
+	}
+}
+
+func TestShow(t *testing.T) {
```
Comment (agis) [inline server_test.go]: Ditto here.

File: server_test.go [hunk 23]
```diff
@@ -65,3 +72,87 @@ func TestLoad(t *testing.T) {
...
+	// Build a job.
```
Comment (agis) [inline server_test.go]: Redundant.

File: types/build_result.go [hunk 24]
```diff
@@ -23,6 +24,9 @@ type BuildResult struct {
 
 	// The method by which the build artifacts can be fetched.
 	TransportMethod TransportMethod
+
+	// The job's start time
+	StartTime time.Time
```
Comment (agis) [inline types/build_result.go]: This belongs in `Job`.

File: types/job_info.go (hunk 5 above)
Comment (agis) [reply types/job_info.go]: Bump.

**Comment by agis** [comment]
Please also add a small section in the README on where the web view can be found.

File: job.go [hunk 25]
```diff
@@ -207,3 +207,18 @@ func (j *Job) String() string {
...
+	_, err := os.Stat(pPath)
```
Comment (agis) [inline job.go]: Instead for checking through the filesystem, which is racy if other methods of mistry manipulate the same paths, you can use the `JobQueue` struct which the server already contains.

**Comment by agis** [comment]
We should add a new issue tracking what we need to revamp in the implementation.

File: types/job_info.go (hunk 5 above)
Comment (apostergiou) [reply types/job_info.go]: Done.

File: README.md (hunk 15 above)
Comment (apostergiou) [reply README.md]: True, fixed.

File: public/templates/show.html (hunk 19 above)
Comment (apostergiou) [reply public/templates/show.html]: done

File: server_test.go (hunk 21 above)
Comment (apostergiou) [reply server_test.go]: Fixed

File: server_test.go (hunk 22 above)
Comment (apostergiou) [reply server_test.go]: Fixed

File: server_test.go (hunk 23 above)
Comment (apostergiou) [reply server_test.go]: Removed

File: types/build_result.go (hunk 24 above)
Comment (apostergiou) [reply types/build_result.go]: Moved to `Job`.

File: job.go (hunk 16 above)
Comment (apostergiou) [reply job.go]: It will return the error.

File: job.go (hunk 25 above)
Comment (apostergiou) [reply job.go]: Good catch, but while I agree that this is racy, what would happen when we restart the server?The `JobQueue` would be empty then and the index page wouldn't display any jobs even if there are some. That's why I believe we should scan the filesystem.

File: job.go (hunk 17 above)
Comment (agis) [reply job.go]: Correction: we do return a non-nil error here, but still, what does an empty string and a "ready path doesn't exist" error means?

File: job.go (hunk 18 above)
Comment (apostergiou) [reply job.go]: I don't think we can as the job has been initialized with an empty struct 
```go
j := &Job{}
state, err := j.GetState(s.cfg.BuildPath, project, id)
```

Probably we can introduce a `GetJob` function that would find a job, WDYT?

File: public/templates/show.html (hunk 20 above)
Comment (agis) [reply public/templates/show.html]: https://developer.mozilla.org/el/docs/Web/JavaScript/Reference/Global_Objects/Error#Examples

File: job.go (hunk 16 above)
Comment (agis) [reply job.go]: You're right, nvm.

File: job.go (hunk 25 above)
Comment (agis) [reply job.go]: Let's revisit this at the next iteration of the web view. It's fine for now.

File: job.go (hunk 17 above)
Comment (apostergiou) [reply job.go]: Thanks for noticing, maybe we could return a custom error then in which we have concatenated any errors with a meaningful message, e.g. "state not found for job %s error %v"

Comment (agis) [reply job.go]: A simple `"job %s not found error", j` I think is sufficient.

File: job.go (hunk 18 above)
Comment (agis) [reply job.go]: Didn't notice that. Jobs should not be initialized this way, we should only use `NewJob()`.

File: job.go (hunk 17 above)
Comment (apostergiou) [reply job.go]: Done

**Comment by apostergiou** [comment]
> We should add a new issue tracking what we need to revamp in the implementation.

Tracked on #38.

**Pull request merged**